File: Rokurics-Windows/Rokurics/CanonicalCore/CanonicalProjectionContract.cs
using System.Globalization;

namespace Rokurics.CanonicalCore;

public static class CanonicalProjectionContract
{
    private const string UnitSeparator = "\u001F";

    public static readonly HashSet<CanonicalArtifactKind> GeneratedArtifactKinds = new()
    {
        CanonicalArtifactKind.TranscriptJson,
        CanonicalArtifactKind.TranscriptMarkdown,
        CanonicalArtifactKind.NoteMarkdown,
        CanonicalArtifactKind.NoteJson,
        CanonicalArtifactKind.SummaryJson
    };

    public static string ArtifactId(string objectId, CanonicalArtifactKind kind) =>
        kind.ArtifactIdFor(NormalizedRequired(objectId, "unknown-recording"));

    public static string ArtifactKey(string objectId, CanonicalArtifactKind kind) =>
        $"{NormalizedRequired(objectId, "unknown-recording")}|{CanonicalArtifactKindRawValue.Of(kind)}";

    public static CanonicalLibraryObjectId MakeCanonicalFolderId(string folderId) =>
        new(folderId, "folder:unknown");

    public static CanonicalLibraryObjectId MakeCanonicalStudyItemId(string itemId) =>
        new(itemId, "studyItem:unknown");

    public static string NormalizeFolderName(string name) =>
        StringUtil.NilIfEmpty(name?.Trim()) ?? "未命名文件夹";

    public static string NormalizeStudyItemTitle(string title, CanonicalStudyItemKind itemKind = CanonicalStudyItemKind.Unknown) =>
        StringUtil.NilIfEmpty(title?.Trim()) ?? (itemKind == CanonicalStudyItemKind.StandaloneNote ? "未命名笔记" : "未命名条目");

    public static List<string> NormalizeTags(List<string>? tags)
    {
        if (tags is null) return new List<string>();
        return new HashSet<string>(
            tags
                .Select(t => StringUtil.NilIfEmpty(t?.Trim())?.ToLowerInvariant())
                .Where(t => t is not null)
                .Cast<string>()
        ).OrderBy(t => t).ToList();
    }

    public static CanonicalHierarchyPath NormalizeFilingPath(CanonicalHierarchyPath path) =>
        new(path.Components);

    public static List<CanonicalParentReference> NormalizeParentReferences(List<CanonicalParentReference>? references)
    {
        if (references is null) return new List<CanonicalParentReference>();
        var seen = new HashSet<string>();
        return references
            .OrderBy(r => r.ParentId.RawValue)
            .Where(r => seen.Add($"{r.Relation}|{r.ParentId.RawValue}"))
            .ToList();
    }

    public static CanonicalTimestamp? NormalizeTombstone(bool isDeleted, CanonicalTimestamp? deletedAt) =>
        isDeleted ? deletedAt : null;

    public static Dictionary<string, string> MetadataHashPayload(CanonicalFolderMetadata folder) =>
        new()
        {
            ["schema"] = "canonical-folder-business-metadata-v1",
            ["folderID"] = folder.FolderId.RawValue,
            ["name"] = folder.Name,
            ["parentID"] = folder.ParentId?.RawValue ?? "",
            ["hierarchyPath"] = folder.HierarchyPath.StableKey,
            ["hierarchyLevel"] = folder.HierarchyLevel ?? "",
            ["colorToken"] = folder.ColorToken ?? "",
            ["orderingKey"] = folder.OrderingKey ?? "",
            ["isDeleted"] = folder.IsDeleted ? "true" : "false",
            ["deletedAt"] = folder.DeletedAt is not null ? TimestampString(folder.DeletedAt) : ""
        };

    public static Dictionary<string, string> MetadataHashPayload(CanonicalStudyItemMetadata item) =>
        new()
        {
            ["schema"] = "canonical-study-item-business-metadata-v1",
            ["itemID"] = item.ItemId.RawValue,
            ["itemKind"] = CanonicalStudyItemKindRawValue.Of(item.ItemKind),
            ["title"] = item.Title,
            ["filingPath"] = item.FilingPath.StableKey,
            ["folderIDs"] = string.Join(UnitSeparator, item.FolderIds.Select(f => f.RawValue)),
            ["parentReferences"] = string.Join(UnitSeparator, item.ParentReferences.Select(p => $"{p.Relation}:{p.ParentId.RawValue}")),
            ["tags"] = string.Join(UnitSeparator, item.Tags),
            ["resourceTokens"] = string.Join(UnitSeparator, item.LogicalResourceTokens),
            ["associatedRecordingID"] = item.AssociatedRecordingId ?? "",
            ["isDeleted"] = item.IsDeleted ? "true" : "false",
            ["deletedAt"] = item.DeletedAt is not null ? TimestampString(item.DeletedAt) : ""
        };

    public static CanonicalObjectKind ObjectKind(string? legacyItemKind, string? recordingId)
    {
        var kind = (legacyItemKind ?? "").Trim();
        if (StringUtil.NilIfEmpty(recordingId?.Trim()) is not null)
            return CanonicalObjectKind.RecordingAssociatedStudyItem;
        if (kind == "standaloneNote")
            return CanonicalObjectKind.StandaloneNote;
        return string.IsNullOrEmpty(kind) ? CanonicalObjectKind.UnknownUnsupported : CanonicalObjectKind.StandaloneStudyItem;
    }

    public static string? SafeLogicalResourceToken(string? token) =>
        SafeLogicalPathToken(token);

    public static string? SafeLogicalPathToken(string? token)
    {
        var trimmed = token?.Trim();
        if (string.IsNullOrEmpty(trimmed)
            || trimmed.StartsWith('/')
            || trimmed.Contains("://")
            || trimmed.Contains('\\'))
            return null;

        var components = trimmed.Split('/', StringSplitOptions.None);
        if (components.Length == 0
            || components.Any(c => string.IsNullOrEmpty(c) || c == "." || c == ".."))
            return null;

        return trimmed;
    }

    public static string? LogicalName(string? token)
    {
        var safeToken = SafeLogicalPathToken(token);
        if (safeToken is null) return null;
        var parts = safeToken.Split('/');
        return parts.Length > 0 ? parts[^1] : null;
    }

    public static CanonicalArtifactProducer Producer(CanonicalArtifactKind kind, string platform)
    {
        var normalizedPlatform = (platform ?? "").Trim().ToLowerInvariant();
        switch (kind)
        {
            case CanonicalArtifactKind.Audio:
                return normalizedPlatform.Contains("iphone") ? CanonicalArtifactProducer.AudioCapture : CanonicalArtifactProducer.Unknown;
            case CanonicalArtifactKind.TranscriptJson:
            case CanonicalArtifactKind.TranscriptMarkdown:
                return normalizedPlatform.Contains("mac") ? CanonicalArtifactProducer.Transcription : CanonicalArtifactProducer.Unknown;
            case CanonicalArtifactKind.NoteMarkdown:
            case CanonicalArtifactKind.NoteJson:
            case CanonicalArtifactKind.SummaryJson:
                return normalizedPlatform.Contains("mac") ? CanonicalArtifactProducer.NoteGeneration : CanonicalArtifactProducer.Unknown;
            default:
                return CanonicalArtifactProducer.Unknown;
        }
    }

    public static CanonicalCapability? RequiredCapability(CanonicalArtifactKind kind) =>
        kind switch
        {
            CanonicalArtifactKind.Audio => CanonicalCapability.AudioArtifact,
            CanonicalArtifactKind.TranscriptJson => CanonicalCapability.TranscriptArtifact,
            CanonicalArtifactKind.TranscriptMarkdown => CanonicalCapability.TranscriptArtifact,
            CanonicalArtifactKind.NoteMarkdown => CanonicalCapability.NoteArtifact,
            CanonicalArtifactKind.NoteJson => CanonicalCapability.NoteArtifact,
            CanonicalArtifactKind.SummaryJson => CanonicalCapability.SummaryArtifact,
            _ => null
        };

    public static CanonicalArtifactAvailability Availability(bool isPresent, CanonicalHash? contentHash, long? byteSize)
    {
        if (!isPresent) return CanonicalArtifactAvailability.Missing;
        return contentHash is not null && byteSize is not null
            ? CanonicalArtifactAvailability.Available
            : CanonicalArtifactAvailability.AvailableWithoutHash;
    }

    public static CanonicalArtifact MakeArtifact(
        string objectId,
        CanonicalArtifactKind kind,
        CanonicalArtifactAvailability availability,
        CanonicalHash? contentHash = null,
        long? byteSize = null,
        string? logicalPathToken = null,
        CanonicalTimestamp? modifiedAt = null,
        CanonicalTimestamp? observedAt = null,
        string? producedByNodeId = null,
        string platform = "")
    {
        var safeToken = SafeLogicalPathToken(logicalPathToken);
        var producer = Producer(kind, platform);
        return new CanonicalArtifact(
            artifactId: ArtifactId(objectId, kind),
            objectId: objectId,
            kind: kind,
            availability: availability,
            contentHash: contentHash,
            byteSize: byteSize,
            logicalName: LogicalName(safeToken),
            logicalPathToken: safeToken,
            modifiedAt: modifiedAt,
            observedAt: observedAt,
            producedBy: producer == CanonicalArtifactProducer.Unknown ? null : producer,
            producedByNodeId: producedByNodeId
        );
    }

    public static bool ProvesGeneratedArtifactAvailability(CanonicalArtifact? artifact)
    {
        if (artifact is null) return false;
        if (!GeneratedArtifactKinds.Contains(artifact.Kind)) return false;
        if (artifact.Availability != CanonicalArtifactAvailability.Available) return false;
        if (artifact.ContentHash is null) return false;
        if (artifact.ByteSize is null) return false;
        if (artifact.Tombstone == true) return false;
        return true;
    }

    public static bool SameContent(CanonicalArtifact left, CanonicalArtifact right) =>
        left.ContentHash?.Algorithm == right.ContentHash?.Algorithm
        && left.ContentHash?.Value == right.ContentHash?.Value
        && left.ByteSize == right.ByteSize
        && left.ContentHash is not null
        && left.ByteSize is not null;

    public static bool IsAuthoritativeProducer(CanonicalArtifact artifact, CanonicalNode node)
    {
        if (artifact.Tombstone == true) return false;
        var requiredCap = RequiredCapability(artifact.Kind);
        if (requiredCap is null) return false;
        if (!node.Capabilities.Contains(requiredCap.Value)) return false;
        if (artifact.ProducedByNodeId is not null && artifact.ProducedByNodeId != node.NodeId)
            return false;

        switch (artifact.Kind)
        {
            case CanonicalArtifactKind.Audio:
                return artifact.ProducedBy == CanonicalArtifactProducer.AudioCapture
                    && node.Platform.ToLowerInvariant().Contains("iphone");
            case CanonicalArtifactKind.TranscriptJson:
            case CanonicalArtifactKind.TranscriptMarkdown:
                return artifact.ProducedBy == CanonicalArtifactProducer.Transcription
                    && node.Platform.ToLowerInvariant().Contains("mac");
            case CanonicalArtifactKind.NoteMarkdown:
            case CanonicalArtifactKind.NoteJson:
            case CanonicalArtifactKind.SummaryJson:
                return artifact.ProducedBy == CanonicalArtifactProducer.NoteGeneration
                    && node.Platform.ToLowerInvariant().Contains("mac");
            default:
                return false;
        }
    }

    internal static string TimestampString(CanonicalTimestamp timestamp)
    {
        var seconds = (timestamp.Date.ToUniversalTime().Ticks - DateTime.UnixEpoch.Ticks) / (double)TimeSpan.TicksPerSecond;
        return seconds.ToString("F6", CultureInfo.InvariantCulture);
    }

    private static string NormalizedRequired(string value, string fallback) =>
        StringUtil.NilIfEmpty(value?.Trim()) ?? fallback;
}

File: Rokurics-Android/app/src/main/java/com/rokurics/app/domain/canonical/CanonicalProjectionContract.kt
package com.rokurics.app.domain.canonical

import java.util.Locale

enum class CanonicalArtifactProducer(val value: String) {
    AUDIO_CAPTURE("audioCapture"),
    TRANSCRIPTION("transcription"),
    NOTE_GENERATION("noteGeneration"),
    UNKNOWN("unknown");
}

object CanonicalProjectionContract {

    private fun normalizedRequired(value: String, fallback: String): String =
        value.trim().nilIfEmpty ?: fallback

    private fun timestampString(timestamp: CanonicalTimestamp): String =
        String.format(Locale.US, "%.6f", timestamp.date.time / 1000.0)

    val generatedArtifactKinds: Set<CanonicalArtifact.Kind> = setOf(
        CanonicalArtifact.Kind.TRANSCRIPT_JSON,
        CanonicalArtifact.Kind.TRANSCRIPT_MARKDOWN,
        CanonicalArtifact.Kind.NOTE_MARKDOWN,
        CanonicalArtifact.Kind.NOTE_JSON,
        CanonicalArtifact.Kind.SUMMARY_JSON
    )

    fun artifactID(objectID: String, kind: CanonicalArtifact.Kind): String =
        kind.artifactID(normalizedRequired(objectID, "unknown-recording"))

    fun artifactKey(objectID: String, kind: CanonicalArtifact.Kind): String =
        "${normalizedRequired(objectID, "unknown-recording")}|${kind.name.lowercase()}"

    fun makeCanonicalFolderID(folderID: String): CanonicalLibraryObjectID =
        CanonicalLibraryObjectID(folderID, "folder:unknown")

    fun makeCanonicalStudyItemID(itemID: String): CanonicalLibraryObjectID =
        CanonicalLibraryObjectID(itemID, "studyItem:unknown")

    fun normalizeFolderName(name: String): String =
        name.trim().nilIfEmpty ?: "未命名文件夹"

    fun normalizeStudyItemTitle(title: String, itemKind: CanonicalStudyItemKind = CanonicalStudyItemKind.UNKNOWN): String =
        title.trim().nilIfEmpty ?: if (itemKind == CanonicalStudyItemKind.STANDALONE_NOTE) "未命名笔记" else "未命名条目"

    fun normalizeTags(tags: List<String>): List<String> =
        tags.mapNotNull { it.trim().nilIfEmpty?.lowercase() }
            .toSet()
            .sorted()

    fun normalizeFilingPath(path: CanonicalHierarchyPath): CanonicalHierarchyPath =
        CanonicalHierarchyPath(path.components)

    fun normalizeParentReferences(references: List<CanonicalParentReference>): List<CanonicalParentReference> {
        val seen = mutableSetOf<String>()
        return references
            .sortedBy { it.parentID.rawValue }
            .filter { reference ->
                val key = "${reference.relation}|${reference.parentID.rawValue}"
                seen.add(key)
            }
    }

    fun normalizeTombstone(isDeleted: Boolean, deletedAt: CanonicalTimestamp?): CanonicalTimestamp? =
        if (isDeleted) deletedAt else null

    fun metadataHashPayload(folder: CanonicalFolderMetadata): Map<String, String> = mapOf(
        "schema" to "canonical-folder-business-metadata-v1",
        "folderID" to folder.folderID.rawValue,
        "name" to folder.name,
        "parentID" to (folder.parentID?.rawValue ?: ""),
        "hierarchyPath" to folder.hierarchyPath.stableKey,
        "hierarchyLevel" to (folder.hierarchyLevel ?: ""),
        "colorToken" to (folder.colorToken ?: ""),
        "orderingKey" to (folder.orderingKey ?: ""),
        "isDeleted" to if (folder.isDeleted) "true" else "false",
        "deletedAt" to (folder.deletedAt?.let { timestampString(it) } ?: "")
    )

    fun metadataHashPayload(item: CanonicalStudyItemMetadata): Map<String, String> = mapOf(
        "schema" to "canonical-study-item-business-metadata-v1",
        "itemID" to item.itemID.rawValue,
        "itemKind" to item.itemKind.value,
        "title" to item.title,
        "filingPath" to item.filingPath.stableKey,
        "folderIDs" to item.folderIDs.joinToString(separator = "\u001F") { it.rawValue },
        "parentReferences" to item.parentReferences.joinToString(separator = "\u001F") { "${it.relation}:${it.parentID.rawValue}" },
        "tags" to item.tags.joinToString(separator = "\u001F"),
        "resourceTokens" to item.logicalResourceTokens.joinToString(separator = "\u001F"),
        "associatedRecordingID" to (item.associatedRecordingID ?: ""),
        "isDeleted" to if (item.isDeleted) "true" else "false",
        "deletedAt" to (item.deletedAt?.let { timestampString(it) } ?: "")
    )

    fun objectKind(
        legacyItemKind: String,
        recordingID: String?
    ): CanonicalObjectKind {
        val kind = legacyItemKind.trim()
        if (recordingID?.trim()?.nilIfEmpty != null) {
            return CanonicalObjectKind.RECORDING_ASSOCIATED_STUDY_ITEM
        }
        if (kind == "standaloneNote") {
            return CanonicalObjectKind.STANDALONE_NOTE
        }
        return if (kind.isEmpty()) CanonicalObjectKind.UNKNOWN_UNSUPPORTED else CanonicalObjectKind.STANDALONE_STUDY_ITEM
    }

    fun safeLogicalResourceToken(token: String?): String? =
        safeLogicalPathToken(token)

    fun safeLogicalPathToken(token: String?): String? {
        val trimmed = token?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        if (trimmed.startsWith("/") || trimmed.contains("://") || trimmed.contains("\\")) return null
        if (trimmed.contains("//") || trimmed.endsWith("/")) return null
        val components = trimmed.split("/")
        if (components.isEmpty() || components.any { it == "." || it == ".." }) return null
        return trimmed
    }

    fun logicalName(token: String?): String? {
        val safeToken = safeLogicalPathToken(token) ?: return null
        return safeToken.split("/").lastOrNull()
    }

    fun producer(kind: CanonicalArtifact.Kind, platform: String): CanonicalArtifactProducer {
        val normalizedPlatform = platform.trim().lowercase()
        return when (kind) {
            CanonicalArtifact.Kind.AUDIO ->
                if (normalizedPlatform.contains("iphone")) CanonicalArtifactProducer.AUDIO_CAPTURE else CanonicalArtifactProducer.UNKNOWN
            CanonicalArtifact.Kind.TRANSCRIPT_JSON, CanonicalArtifact.Kind.TRANSCRIPT_MARKDOWN ->
                if (normalizedPlatform.contains("mac")) CanonicalArtifactProducer.TRANSCRIPTION else CanonicalArtifactProducer.UNKNOWN
            CanonicalArtifact.Kind.NOTE_MARKDOWN, CanonicalArtifact.Kind.NOTE_JSON, CanonicalArtifact.Kind.SUMMARY_JSON ->
                if (normalizedPlatform.contains("mac")) CanonicalArtifactProducer.NOTE_GENERATION else CanonicalArtifactProducer.UNKNOWN
            CanonicalArtifact.Kind.METADATA, CanonicalArtifact.Kind.RECEIVE_RECORD ->
                CanonicalArtifactProducer.UNKNOWN
        }
    }

    fun requiredCapability(kind: CanonicalArtifact.Kind): CanonicalCapability? {
        return when (kind) {
            CanonicalArtifact.Kind.AUDIO -> CanonicalCapability.AUDIO_ARTIFACT
            CanonicalArtifact.Kind.TRANSCRIPT_JSON, CanonicalArtifact.Kind.TRANSCRIPT_MARKDOWN -> CanonicalCapability.TRANSCRIPT_ARTIFACT
            CanonicalArtifact.Kind.NOTE_MARKDOWN, CanonicalArtifact.Kind.NOTE_JSON -> CanonicalCapability.NOTE_ARTIFACT
            CanonicalArtifact.Kind.SUMMARY_JSON -> CanonicalCapability.SUMMARY_ARTIFACT
            CanonicalArtifact.Kind.METADATA, CanonicalArtifact.Kind.RECEIVE_RECORD -> null
        }
    }

    fun availability(
        isPresent: Boolean,
        contentHash: CanonicalHash?,
        byteSize: Long?
    ): CanonicalArtifact.Availability {
        if (!isPresent) return CanonicalArtifact.Availability.MISSING
        return if (contentHash != null && byteSize != null)
            CanonicalArtifact.Availability.AVAILABLE
        else
            CanonicalArtifact.Availability.AVAILABLE_WITHOUT_HASH
    }

    fun makeArtifact(
        objectID: String,
        kind: CanonicalArtifact.Kind,
        availability: CanonicalArtifact.Availability,
        contentHash: CanonicalHash? = null,
        byteSize: Long? = null,
        logicalPathToken: String? = null,
        modifiedAt: CanonicalTimestamp? = null,
        observedAt: CanonicalTimestamp? = null,
        producedByNodeID: String? = null,
        platform: String
    ): CanonicalArtifact {
        val safeToken = safeLogicalPathToken(logicalPathToken)
        val producer = producer(kind, platform)
        return CanonicalArtifact(
            artifactID = artifactID(objectID, kind),
            objectID = objectID,
            kind = kind,
            availability = availability,
            contentHash = contentHash,
            byteSize = byteSize,
            logicalName = logicalName(safeToken),
            logicalPathToken = safeToken,
            modifiedAt = modifiedAt,
            observedAt = observedAt,
            producedBy = if (producer == CanonicalArtifactProducer.UNKNOWN) null else producer,
            producedByNodeID = producedByNodeID,
            tombstone = null
        )
    }

    fun provesGeneratedArtifactAvailability(artifact: CanonicalArtifact?): Boolean {
        val a = artifact ?: return false
        return generatedArtifactKinds.contains(a.kind) &&
            a.availability == CanonicalArtifact.Availability.AVAILABLE &&
            a.contentHash != null &&
            a.byteSize != null &&
            a.tombstone != true
    }

    fun sameContent(left: CanonicalArtifact, right: CanonicalArtifact): Boolean {
        return left.contentHash?.algorithm == right.contentHash?.algorithm &&
            left.contentHash?.value == right.contentHash?.value &&
            left.byteSize == right.byteSize &&
            left.contentHash != null &&
            left.byteSize != null
    }

    fun isAuthoritativeProducer(artifact: CanonicalArtifact, node: CanonicalNode): Boolean {
        if (artifact.tombstone == true) return false
        val requiredCap = requiredCapability(artifact.kind) ?: return false
        if (!node.capabilities.contains(requiredCap)) return false
        if (artifact.producedByNodeID != null && artifact.producedByNodeID != node.nodeID) return false
        return when (artifact.kind) {
            CanonicalArtifact.Kind.AUDIO ->
                artifact.producedBy == CanonicalArtifactProducer.AUDIO_CAPTURE &&
                    node.platform.lowercase().contains("iphone")
            CanonicalArtifact.Kind.TRANSCRIPT_JSON, CanonicalArtifact.Kind.TRANSCRIPT_MARKDOWN ->
                artifact.producedBy == CanonicalArtifactProducer.TRANSCRIPTION &&
                    node.platform.lowercase().contains("mac")
            CanonicalArtifact.Kind.NOTE_MARKDOWN, CanonicalArtifact.Kind.NOTE_JSON, CanonicalArtifact.Kind.SUMMARY_JSON ->
                artifact.producedBy == CanonicalArtifactProducer.NOTE_GENERATION &&
                    node.platform.lowercase().contains("mac")
            CanonicalArtifact.Kind.METADATA, CanonicalArtifact.Kind.RECEIVE_RECORD -> false
        }
    }
}
